android/common/src/main/java/com/solana/mobilewalletadapter/common/protocol/MobileWalletAdapterSessionCommon.java

```diff
@@ -195,13 +205,20 @@ protected byte[] decryptSessionPayload(@NonNull byte[] payload) throws SessionMe
             throw new IllegalStateException("Cannot decrypt, no session key has been established");
         }
 
+        final int seqNum = ByteBuffer.wrap(payload, 0, SEQ_NUM_LENGTH_BYTES).getInt(); // Big-endian
+        if (seqNum <= mSeqNumberRx) {
```

Comment:
This would allow skipping a sequence number. Do we want to allow that?

Reply:
If yes, cool! If not, then the variable we should track is more like `mNextSeqNumberRx` and then do…

```
if (seqNum != mNextSeqNumberRx) {
  throw...
}
mNextSeqNumberRx++;
```

Reply:
I spent some time thinking about whether to shoot for strict sequence numbers, or permissive sequence numbers. I ended up deciding on permissive sequence numbers, because the spec is designed to support multiple different transports, and it doesn't set any hard requirements on whether or not a transport must be lossless (a la Websockets over TCP). Monotonically increasing sequence numbers hardens against replay attacks, but also allows for transports that:
- may fail in a way that the sender can observe. While we could roll back the sequence number in this case, it adds complexity and inter-module coupling to the sender's encryption implementation
- may fail in a way that the sender cannot observe. In this case, while one operation may fail, the channel remains resilient. I debated whether this is desirable behavior, since a failure to send implies a failure to provide service, but ultimately decided that this is already a property of a lossy transport, and adding sequence numbers can't fix this, so sequence numbers should be tolerant of it (for any application of the protocol that chooses to use a lossy transport method).